LUCENE-9330: Make SortFields responsible for index sorting and serialization

romseygeek opened the pull request:
This commit adds a new class `IndexSorter` which handles how a sort should be applied
to documents in an index:
* how to serialize/deserialize sort info in the segment header
* how to sort documents within a segment
* how to sort documents from merging segments

SortField has a `getIndexSorter()` method, which will return `null` if the sort cannot be used
to sort an index (eg if it uses scores or other query-dependent values).  This also requires a
new Codec as there is a change to the SegmentInfoFormat

romseygeek commented:
This is quite a big PR, but most of it is around the creation of a new Lucene86Codec.  The important bit to review are the changes to DefaultIndexingChain and the DocValuesWriters, and the implementation of IndexSorter for existing sort fields (int, long, float, double, string, sortednumeric, sortedset).

uschindler commented:
I left a comment on the issue about the serialization of class names: Don't do this, will break with Java's module system. We should use NamedSPI for that, like PostingsFormat or Codecs.

uschindler commented:
Using SPI makes it much better now! Thanks.

jpountz commented on lucene/backward-codecs/src/java/org/apache/lucene/codecs/lucene84/package.html:
```diff
@@ -0,0 +1,42 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
```
Two licenses?

jpountz commented on lucene/codecs/src/java/org/apache/lucene/codecs/simpletext/SimpleTextSegmentInfoFormat.java:
```diff
@@ -171,133 +168,17 @@ public SegmentInfo read(Directory directory, String segmentName, byte[] segmentI
+        assert StringHelper.startsWith(scratch.get(), SI_SORT_BYTES);
+        BytesRef serializedSort = SimpleTextUtil.fromBytesRefString(readString(SI_SORT_BYTES.length, scratch));
+        final ByteArrayDataInput bytes = new ByteArrayDataInput(serializedSort.bytes, serializedSort.offset, serializedSort.length);
+        sortField[i] = SortFieldProvider.forName(provider).loadSortField(bytes);
```
assert that the bytes have been fully consumed?

jpountz commented on lucene/core/src/java/org/apache/lucene/index/DocValuesWriter.java:
```diff
@@ -21,12 +21,10 @@
   abstract void flush(SegmentWriteState state, Sorter.DocMap sortMap, DocValuesConsumer consumer) throws IOException;
-  abstract Sorter.DocComparator getDocComparator(int numDoc, SortField sortField) throws IOException;
+  abstract T getDocValues();
   abstract DocIdSetIterator getDocIdSet();
```
can we remove this since `getDocValues` already returns an iterator? (we might need to do `T extends DocIdSetIterator` above)

jpountz commented on lucene/core/src/java/org/apache/lucene/index/IndexSorter.java:
```diff
@@ -0,0 +1,500 @@
+   *
+   * @see SortFieldProvider#loadSortField(DataInput)
+   */
+  public abstract void serialize(DataOutput out) throws IOException;
```
I wonder if it would work to move it to SortFieldProvider instead for symmetry, so that the serialization and deserialization logic would be in the same place.

jpountz commented on lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:
```diff
@@ -94,29 +89,100 @@ public DefaultIndexingChain(DocumentsWriterPerThread docWriter) {
     Sorter sorter = new Sorter(indexSort);
     // returns null if the documents are already sorted
-    return sorter.sort(state.segmentInfo.maxDoc(), comparators.toArray(new Sorter.DocComparator[comparators.size()]));
+    return sorter.sort(state.segmentInfo.maxDoc(), comparators.toArray(new IndexSorter.DocComparator[0]));
```
There is an even nicer syntax now.

```suggestion
    return sorter.sort(state.segmentInfo.maxDoc(), comparators.toArray(IndexSorter.DocComparator[]::new));
```

jpountz commented on lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:
```diff
@@ -527,45 +589,61 @@ private void indexPoint(PerField fp, IndexableField field) throws IOException {
-            }
-            break;
+      IndexSorter sorter = sortField.getIndexSorter();
+      assert sorter != null;
```
should it be an actual exception instead of an assertion? Or are we checking it up-front already?

romseygeek replied:
It's checked in `IndexWriterConfig.setSort()` but I'll add an exception here as well; you could conceivably get here via a funky Codec that instantiates a Sort with no IndexSorter defined.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/SortField.java:
```diff
@@ -392,4 +491,31 @@ public SortField rewrite(IndexSearcher searcher) throws IOException {
+   * SortFields that implement this method should also implement a constructor that
+   * takes a {@link DataInput} for deserialization, to match the {@link IndexSorter#serialize(DataOutput)}
+   * method on the returned IndexSorter
+   */
```
make it `lucene.internal`?

romseygeek replied:
I'm not sure this API is particularly internal? If people want to implement their own SortFields and make them indexable they will need to implement this method, which makes it public IMO.

We should probably take `@lucene.experimental` off `getFieldComparator()` as well, it's been there since 2004...

jpountz replied:
I suggested it because I was seeing your change as a cleanup to the way that we support configuring index sorts. experimental would work for me too, what I like is to signal users that we feel free to make breaking changes to these APIs.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/SortField.java:
```diff
@@ -120,6 +126,99 @@ public SortField(String field, Type type, boolean reverse) {
+  public static final class Provider extends SortFieldProvider {
+
+    /** The name this Provider is registered under */
+    public static final String NAME = "field";
```
maybe reuse the class name like we do for codecs, ie. `SortField`?

jpountz commented on lucene/core/src/java/org/apache/lucene/search/SortedNumericSortField.java:
```diff
@@ -83,6 +89,80 @@ public SortedNumericSortField(String field, SortField.Type type, boolean reverse
+  public static final class Provider extends SortFieldProvider {
+
+    /** The name this provider is registered under */
+    public static final String NAME = "sortedNumericField";
```
maybe reuse the class name?

jpountz commented on lucene/core/src/java/org/apache/lucene/search/SortedSetSortField.java:
```diff
@@ -68,6 +72,54 @@ public SortedSetSortField(String field, boolean reverse, SortedSetSelector.Type
+  public static final class Provider extends SortFieldProvider {
+
+    /** The name this provider is registered under */
+    public static final String NAME = "sortedSetField";
```
use the class name?

jpountz commented on lucene/core/src/java/org/apache/lucene/index/IndexSorter.java:
```diff
@@ -0,0 +1,500 @@
+  /**
+   * Sorts documents based on integer values from a NumericDocValues instance
+   */
+  public static final class IntSorter implements IndexSorter {
```
can it be package private?

romseygeek replied:
IndexSorter is an interface (started as an abstract class in a first iteration) so everything is public and static, and we don't need those modifiers anymore.

romseygeek commented:
Thanks for the review @jpountz!  I addressed your feedback, including moving the serialization logic completely into SortFieldProvider.  I'm not entirely happy with it as it requires some hacky casting, but I'm not sure there's a nicer way of doing it.

jpountz commented on lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:
```diff
@@ -94,29 +89,100 @@ public DefaultIndexingChain(DocumentsWriterPerThread docWriter) {
     termsHash = new FreqProxTermsWriter(docWriter, termVectorsWriter);
   }
 
+  private LeafReader getDocValuesReader(int maxDoc) {
```
```suggestion
  private LeafReader getDocValuesLeafReader(int maxDoc) {
```

jpountz commented on lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java:
```diff
@@ -527,45 +589,63 @@ private void indexPoint(PerField fp, IndexableField field) throws IOException {
+
+        @Override
+        public int maxDoc() {
+          return 0;
```
I'd have a slight preferenrece for throwing an UOE here, would it work?

romseygeek replied:
`IndexSorter.getDocComparator(Reader)` calls `maxDoc()` on the reader to allocate its comparison arrays.  We could change the signature so that it takes `maxDoc` as well as the Reader, which would reduce the indirection a bit and make it easier to read?

jpountz replied:
+1

jpountz commented on lucene/core/src/java/org/apache/lucene/index/SortFieldProvider.java:
```diff
@@ -0,0 +1,110 @@
+    Holder.getLoader().reload(classLoader);
+  }
+
+  public static void serialize(SortField sf, DataOutput output) throws IOException {
```
javadocs?

jpountz commented on lucene/core/src/java/org/apache/lucene/index/SortFieldProvider.java:
```diff
@@ -0,0 +1,110 @@
+   */
+  public abstract SortField loadSortField(DataInput in) throws IOException;
+
+  public abstract void writeSortField(SortField sf, DataOutput out) throws IOException;
```
javadocs?

jpountz commented on lucene/core/src/java/org/apache/lucene/index/SortedDocValuesWriter.java:
```diff
@@ -79,11 +78,6 @@ public void addValue(int docID, BytesRef value) {
-  @Override
-  public void finish(int maxDoc) {
-    updateBytesUsed();
-  }
```
I wonder if there are undesired consequences of dropping this?

romseygeek replied:
It was always called either immediately before `flush` or `getDocComparator`, so it seemed to make sense to just fold it directly into those methods.

romseygeek commented:
> Should we remove write support from Lucene70SegmentInfoFormat and have a RW version under
test-framework like we do for other components, so that users can't use it in their codecs but we can still run the segment info format test case?

This is trickier for the segment info format because both reading and writing are handled by the same class.  I think so far we've only done this for Postings formats, and not for other parts of the Codec?

jpountz commented:
> This is trickier for the segment info format because both reading and writing are handled by the same class. I think so far we've only done this for Postings formats, and not for other parts of the Codec?

We've done it in the past already, see e.g. https://github.com/apache/lucene-solr/commit/23b002a0fdf2f6025f1eb026c0afca247fb21ed0. LuceneXXSegmentInfoFormat is changed to throw an UOE in the `write` method, then a LuceneXXRWSegmentInfoFormat is created that extends LuceneXXSegmentInfoFormat and adds back the `write` implementation.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/SortField.java:
```diff
@@ -120,6 +126,104 @@ public SortField(String field, Type type, boolean reverse) {
+
+  private void serialize(DataOutput out) throws IOException {
+    out.writeString(field);
+    out.writeInt(type.ordinal());
```
Can we look into not using enum ordinals for serialization? It makes something that looks as harmless as reordering an enum cause backward-compatibility breaks. E.g. maybe we could have a mapping between Type constants and int IDs as a constant on this class?

jpountz commented on lucene/core/src/java/org/apache/lucene/search/SortedNumericSortField.java:
```diff
@@ -83,6 +89,86 @@ public SortedNumericSortField(String field, SortField.Type type, boolean reverse
+    if (selectorType >= SortedNumericSelector.Type.values().length) {
+      throw new IllegalArgumentException("Can't deserialize SortedNumericSortField - unknown selector type " + selectorType);
+    }
+    return SortedNumericSelector.Type.values()[selectorType];
```
same here?
